isis/src/control/apps/cnetdiff/cnetdiff.cpp

```diff
@@ -0,0 +1,380 @@
+For more details about the LICENSE terms and the AUTHORS, you will
+find files of those names at the top level of this repository. **/
+
+/* SPDX-License-Identifier: CC0-1.0 */
+
+#include "cnetdiff.h"
+
+#include <cmath>
+#include <float.h>
+
+#include <QList>
+#include <QString>
+
+#include "ControlNet.h"
+#include "ControlNetDiff.h"
+#include "ControlNetVersioner.h"
+#include "IException.h"
+#include "PvlContainer.h"
+#include "PvlGroup.h"
+
+using namespace std;
+
+namespace Isis {
+  
+  bool filesMatch;
+  QString differenceReason;
+  PvlGroup tolerances;
+  PvlGroup ignorekeys;
+
+  void CompareKeywords(const PvlKeyword &pvl1, const PvlKeyword &pvl2);
+  void CompareGroups(const PvlContainer &pvl1, const PvlContainer &pvl2);
+  void Compare(const PvlObject &point1, const PvlObject &point2);
+  void Compare(ControlNetVersioner &net1, ControlNetVersioner &net2);
+
+  /**
+   * Compare two control networks
+   *
+   * @param ui UserInterface object containing parameters
+   * @return Pvl results log file
+   */
+  Pvl cnetdiff(UserInterface &ui) {
+
+    // input control networks
+    ControlNet cnet1 = ControlNet(ui.GetFileName("FROM"));
+    ControlNet cnet2 = ControlNet(ui.GetFileName("FROM2"));
+
+    if(ui.WasEntered("DIFF")) {
+      Pvl diffFile(ui.GetFileName("DIFF"));
+      return cnetdiff(cnet1, cnet2, ui, &diffFile);
+    }
+    
+    return cnetdiff(cnet1, cnet2, ui);
+  }
+
+
+  /**
+   * Compare two control networks
+   *
+   * @param ControlNet cnet1 1st control net for comparison
+   * @param ControlNet cnet2 2nd control net for comparison
+   * @return Pvl results log file
+   */
+  Pvl cnetdiff(ControlNet &cnet1, ControlNet &cnet2, UserInterface &ui,
+               Pvl *diffFile) {
+    Pvl log;
+
+    // create ControlNetVersioner objects for each net
+    ControlNetVersioner cnv1(&cnet1);
+    ControlNetVersioner cnv2(&cnet2);
+
+    // report first difference only
+    if (ui.GetString("REPORT") == "FIRST") {
+      tolerances = PvlGroup();
+      ignorekeys = PvlGroup();
+
+      differenceReason = "";
+      filesMatch = true;
+
+      if (diffFile != nullptr) {
+        if(diffFile->hasGroup("Tolerances")) {
+          tolerances = diffFile->findGroup("Tolerances");
+        }
+
+        if(diffFile->hasGroup("IgnoreKeys")) {
+          ignorekeys = diffFile->findGroup("IgnoreKeys");
+        }
+      }
+
+      // Don't want to consider the DateTime of a Point or Measure was set by
+      // default.
+      if(!ignorekeys.hasKeyword("DateTime")) {
+        ignorekeys += PvlKeyword("DateTime", "true");
+      }
+
+      // compare ControlNetVersioners
+      Compare(cnv1, cnv2);
+    
+      PvlGroup differences("Results");
+      if (filesMatch) {
+        differences += PvlKeyword("Compare", "Identical");
+      }
+      else {
+        differences += PvlKeyword("Compare", "Different");
+        differences += PvlKeyword("Reason", differenceReason);
+      }
+
+      log.addGroup(differences);
```

Comment:
Could any addition to the log be changed to `log.addLogGroup`. This function logs to the app at the same time as the group is added to the keyword. With that you can remove the logging loop in the application `main.cpp`

Reply:
Also done.